src/preprocessing/preprocessing_pass_context.h

```diff
@@ -79,7 +79,11 @@ class PreprocessingPassContext : protected EnvObj
   /** Spend resource in the resource manager of the associated Env. */
   void spendResource(Resource r);
 
-  /** Get a reference to the top-level substitution map */
+  /**
+   * Get a reference to the top-level substitution map. Note that all
+   * substitutions added to this map should use the addSubstitution methods
+   * below for the purposes of proper debugging information.
+   */
```

Comment:
Given this description, I would expect this method to return a `const &`. I'm aware that this is not trivial to do, but I think we should consider aiming for it...

Reply:
It is hard to do this (I abandoned an attempt on this branch).  This is because it keeps a cache.

Reply:
Yes, I know. I think we should consider making the cache in `SubstitutionMap` mutable, and then make `SubstitutionMap::apply()` const. But this is not in scope for this PR...